### ISSUE TITLE
Apex Does Not Accelerate Training Speed

### ISSUE THREAD

**Issue by researchlab36** [NVIDIA/NeMo#15481]
## Describe the bug

When configuring and running NeMo ASR training with the `fused_adam` optimizer (provided by NVIDIA Apex), the optimizer is correctly loaded and instantiated, however **no measurable training speed improvement** is observed compared to standard optimizer usage.

Debug print statements placed inside the `FusedAdam` / `FusedLAMB` classes confirm that the Apex optimizer class is instantiated and executed during training.

Despite the following being true:

- Apex is successfully imported
- `HAVE_APEX = True`
- `fused_adam` appears in `AVAILABLE_OPTIMIZERS`
- The optimizer class is called during training

there is **no observable improvement in training throughput** (steps/sec or samples/sec) compared to standard PyTorch `Adam`.

Since Apex fused optimizers are designed to reduce optimizer step overhead via fused CUDA kernels, a performance improvement was expected.

---

### Training Script

```python
from omegaconf import OmegaConf
from nemo.utils.exp_manager import exp_manager
import torch
from nemo.collections.asr.models import EncDecCTCModelBPE
from lightning.pytorch import Trainer
import os

os.environ['TMPDIR'] = '/tmp_dir/audio_data/multilingual_asr_nemo/temp_checkpoints'
os.makedirs(os.environ['TMPDIR'], exist_ok=True)

cfg = OmegaConf.load(
    "/path_to_yaml/audio/audio_data/multilingual_asr_nemo/configs/fastconf_long_150_128_apex.yaml"
)

cfg = OmegaConf.create(OmegaConf.to_container(cfg, resolve=True))

trainer = Trainer(**cfg.trainer)
exp_manager(trainer, cfg.get("exp_manager", None))

asr_model = EncDecCTCModelBPE(cfg=cfg.model, trainer=trainer)

trainer.fit(asr_model)
```

---

### Relevant Optimizer Registration (NeMo `optimizers.py`)

```python
try:
    from apex.optimizers import FusedAdam, FusedLAMB

    HAVE_APEX = True
    AVAILABLE_OPTIMIZERS['lamb'] = FusedLAMB
    AVAILABLE_OPTIMIZERS['fused_adam'] = FusedAdam
except Exception:
    HAVE_APEX = False
```

---

### Trainer configuration

```yaml
trainer:
  devices: -1
  num_nodes: 1
  max_epochs: 1000
  max_steps: -1
  val_check_interval: 1.0
  accelerator: auto
  strategy: ddp
  accumulate_grad_batches: 1
  gradient_clip_val: 0.0
  precision: bf16-mixed
  log_every_n_steps: 5000
  enable_progress_bar: true
  num_sanity_val_steps: 0
  check_val_every_n_epoch: 1
  sync_batchnorm: true
  enable_checkpointing: false
  logger: false
  benchmark: false
```

---

### Optimizer configuration

```yaml
optim:
  name: fused_adam
  lr: 1e-3
  betas: [0.9, 0.98]
  weight_decay: 1e-3
  sched:
    name: CosineAnnealing
    warmup_steps: 15000
    warmup_ratio: null
    min_lr: 1e-4
```

---

### Full Model Configuration

```yaml
name: "fastconf_long_150_128_combined_ctc_bpe_interctc_apex"

model:
  sample_rate: 16000
  log_prediction: false
  ctc_reduction: "mean_volume"
  skip_nan_grad: true

  train_ds:
    manifest_filepath: json_path
    sample_rate: 16000
    batch_size: 64
    shuffle: true
    num_workers: 16
    pin_memory: true
    max_duration: 25.0
    min_duration: 0.1
    labels: null
    is_tarred: false
    tarred_audio_filepaths: null
    shuffle_n: 2048
    bucketing_strategy: "synced_randomized"
    bucketing_batch_size: null

  validation_ds:
    manifest_filepath: json_path
    sample_rate: 16000
    labels: null
    batch_size: 6
    shuffle: false
    use_start_end_token: false
    num_workers: 8
    pin_memory: true

  tokenizer:
    type: agg
    langs:
      ka:
        type: bpe
        dir: /tokenizer_path/tokenizer_spe_bpe_v128
      en:
        type: bpe
        dir: /tokenizer_path/tokenizer_spe_bpe_v128

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: 16000
    normalize: per_feature
    window_size: 0.025
    window_stride: 0.01
    window: hann
    features: 80
    n_fft: 512
    log: true
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0
    pad_value: 0.0

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2
    time_masks: 10
    freq_width: 27
    time_width: 0.05

  encoder:
    _target_: nemo.collections.asr.modules.ConformerEncoder
    feat_in: 80
    feat_out: -1
    n_layers: 17
    d_model: 512
    use_bias: true
    subsampling: dw_striding
    subsampling_factor: 8
    subsampling_conv_channels: 256
    causal_downsampling: false
    ff_expansion_factor: 4
    self_attention_model: rel_pos_local_attn
    global_tokens: 1
    global_tokens_spacing: 1
    global_attn_separate: false
    n_heads: 8
    att_context_size: [128, 128]
    att_context_style: regular
    xscaling: true
    untie_biases: true
    conv_kernel_size: 9
    conv_norm_type: batch_norm
    conv_context_size: null
    dropout: 0.1
    dropout_pre_encoder: 0.1
    dropout_emb: 0.0
    dropout_att: 0.1
    stochastic_depth_drop_prob: 0.0
    stochastic_depth_mode: linear
    stochastic_depth_start_layer: 1

  decoder:
    _target_: nemo.collections.asr.modules.ConvASRDecoder
    feat_in: null
    num_classes: -1
    vocabulary: []

  interctc:
    loss_weights: [0.3]
    apply_at_layers: [7]

  optim:
    name: fused_adam
    lr: 1e-3
    betas: [0.9, 0.98]
    weight_decay: 1e-3
    sched:
      name: CosineAnnealing
      warmup_steps: 15000
      warmup_ratio: null
      min_lr: 1e-4
```

---

## Apex Installation

Apex was installed from the official NVIDIA repository:

https://github.com/NVIDIA/apex

Installation command used:

```bash
NVCC_APPEND_FLAGS="--threads 4" \
APEX_PARALLEL_BUILD=8 \
APEX_CPP_EXT=1 \
APEX_CUDA_EXT=1 \
pip install -v --no-build-isolation .
```

---

## Apex Extensions Built

The following Apex CUDA extensions were successfully compiled during installation:

| Extension | Purpose | Needed for ASR |
|---|---|---|
| amp_C | Automatic Mixed Precision | Yes |
| fused_layer_norm_cuda | Fused Layer Norm | Yes |
| syncbn | Synchronized BatchNorm | Yes |
| fused_lamb_cuda | LAMB Optimizer | Yes |
| apex_C | Core utilities | Yes |
| mlp_cuda | Fused MLP | Yes |
| distributed_adam_cuda | Distributed Adam | Yes |
| distributed_lamb_cuda | Distributed LAMB | Yes |
| scaled_masked_softmax_cuda | Scaled Softmax | Yes |
| fused_layer_norm_cuda | Layer Normalization | Yes |

---

## Expected behavior

When using `fused_adam` (Apex `FusedAdam`) as the optimizer, a **measurable improvement in training throughput** (steps/sec or samples/sec) is expected compared to standard PyTorch `Adam`.

Apex fused CUDA kernels should reduce optimizer step overhead and improve performance on NVIDIA GPUs.

---

### Hardware

| Parameter | Value |
|-----------|------|
| GPU Model | 2× NVIDIA A100 80GB |
| GPU Count | 2 |
| Driver | 535.154.05 |
| CUDA | 12.2 |

### Software

| Component | Version |
|-----------|--------|
| Python | 3.10.20 |
| PyTorch | 2.9.1 |
| NeMo | 2.7.0 |
| NVIDIA Driver | 535.154.05 |
| CUDA | 12.2 |

---

## Additional context

- Apex is confirmed to be installed and active (`HAVE_APEX = True`)
- `fused_adam` and `lamb` are successfully registered in `AVAILABLE_OPTIMIZERS`
- Debug print statements confirm that the Apex optimizer class is instantiated
- Training runs using **DDP across 2× A100 GPUs**
- Precision mode: **bf16-mixed**
- Model architecture: **Fast Conformer Long CTC (~120M parameters)** with **InterCTC**
- Tokenizer: **aggregate multilingual BPE (Georgian + English)** with **128 vocab each**

**Comment by pzelasko** [comment]
PyTorch added a `fused=True` flag to AdamW starting in 2.0, I think it's set by default. Basically Apex's optimizations for Adam are available off the shelf for some time now.
